src/browser/file-recovery-service.js

```diff
@@ -0,0 +1,65 @@
+'use babel'
+
+import {ipcMain} from 'electron'
+import crypto from 'crypto'
+import Path from 'path'
+import fs from 'fs-plus'
+
+export default class FileRecoveryService {
+  constructor (recoveryDirectory) {
+    this.recoveryDirectory = recoveryDirectory
+    this.recoveryPathsByWindowAndFilePath = new WeakMap()
+    this.crashListeners = new WeakSet()
+  }
+
+  start () {
+    ipcMain.on('will-save-path', this.willSavePath.bind(this))
+    ipcMain.on('did-save-path', this.didSavePath.bind(this))
+  }
+
+  willSavePath (event, path) {
+    if (!fs.existsSync(path)) {
+      // Unexisting files won't be truncated/overwritten, and so there's no data to be lost.
+      return
+    }
+
+    const window = event.sender
+    const recoveryFileName = crypto.randomBytes(5).toString('hex')
+    const recoveryPath = Path.join(this.recoveryDirectory, recoveryFileName)
+    fs.writeFileSync(recoveryPath, fs.readFileSync(path))
```

Comment:
> Would it be worth using the original file basename as part of the filename so that they can be more easily identified?

My main concern there is path length on Windows, which is why I went with a 10-letter random filename for these recovery files. Introducing a variable-length file name could cause weird edge cases on Windows, although it's admittedly more user-friendly. Maybe we can truncate the filename at some arbitrary length, and add a random, fixed-size suffix string? I'll try that shortly.